services/subscription.go

```diff
@@ -0,0 +1,285 @@
+package services
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/smartcontractkit/chainlink/logger"
+	"github.com/smartcontractkit/chainlink/store"
+	"github.com/smartcontractkit/chainlink/store/models"
+	"github.com/smartcontractkit/chainlink/store/presenters"
+	"github.com/smartcontractkit/chainlink/utils"
+	"go.uber.org/multierr"
+)
+
+// Descriptive indices of a RunLog's Topic array
+const (
+	EventTopicSignature = iota
+	EventTopicRequestID
+	EventTopicJobID
+)
+
+// RunLogTopic is the signature for the Request(uint256,bytes32,string) event
+// which Chainlink RunLog initiators watch for.
+// See https://github.com/smartcontractkit/chainlink/blob/master/solidity/contracts/Oracle.sol
+var RunLogTopic = common.HexToHash("0x06f4bf36b4e011a5c499cef1113c2d166800ce4013f6c2509cab1a0e92b83fb2")
+
+// Listens to event logs being pushed from the Ethereum Node specific to a job.
+type JobSubscription struct {
+	Job                    models.Job
+	InitiatorSubscriptions []InitiatorSubscription
+}
+
+// Constructor of JobSubscription that to starts listening to and keeps track of
+// event logs corresponding to a job.
+func StartJobSubscription(job models.Job, store *store.Store) (JobSubscription, error) {
+	var merr error
+	var initSubs []InitiatorSubscription
+	for _, initr := range job.InitiatorsFor(models.InitiatorEthLog) {
+		sub, err := StartEthLogSubscription(initr, job, store)
+		merr = multierr.Append(merr, err)
+		if err == nil {
+			initSubs = append(initSubs, sub)
+		}
+	}
+
+	for _, initr := range job.InitiatorsFor(models.InitiatorRunLog) {
+		sub, err := StartRunLogSubscription(initr, job, store)
+		merr = multierr.Append(merr, err)
+		if err == nil {
+			initSubs = append(initSubs, sub)
+		}
+	}
+
+	if len(initSubs) == 0 {
+		return JobSubscription{}, multierr.Append(merr, errors.New("Job must have a valid log initiator"))
+	}
+
+	js := JobSubscription{Job: job, InitiatorSubscriptions: initSubs}
+	return js, merr
+}
+
+// Stops the subscription and cleans up associated resources.
+func (js JobSubscription) Unsubscribe() {
+	for _, sub := range js.InitiatorSubscriptions {
+		sub.Unsubscribe()
+	}
+}
+
+// Interface for all subscriptions made specific to a subscription.
+type InitiatorSubscription interface {
+	Unsubscribe()
+}
+
+// Encapsulates all functionality needed to wrap an ethereum rpc.ClientSubscription
+// for use with a Chainlink Initiator. Initiator specific functionality is delegated
+// to the ReceiveLog callback using a strategy pattern.
+type RpcSubscription struct {
```

Comment:
Isn't NewHeads also an RPC subscription? Maybe this is a log subscription?

Reply:
It is indeed. They both return `*rpc.ClientSubscription` from the geth library. This becomes a log subscription because of the callback field `ReceiveLog`. I can rename this type if you would like, or perhaps on a subsequent commit, have `newHeads` use a lower version of this that doesn't force `ReceiveLog(InitiatorLogEvent)`.

I would like to go for the low hanging fruit here so we can merge into master sooner rather than later. Let me know what you think. Maybe it's just a rename to `RpcLogSubscription`